### ISSUE TITLE
default_worker.py  crashes when initializing --storage in ray head node

### ISSUE THREAD

**Issue by tvildo** [ray-project/ray#41568]
### What happened + What you expected to happen

Hello, 
I am trying to start ray head with following parameters:
`ray start --head  --dashboard-host=0.0.0.0 --block --storage="s3://explore/ray?&endpoint_override=http://192.168.0.4:9000"`

When I am trying to start serve with following code:

**main.py**
```
import ray
import time
from ray import serve
import logging

runtime_env = {
    "env_vars": {"TF_WARNINGS": "none"}
}

rayContext = ray.init(address='ray://ray_head_address_goes_here:10001', 
                      runtime_env=runtime_env,
                      logging_level=logging.DEBUG, log_to_driver=True
                    )

print('Starting Serve...')
serve.start(detached=True, http_options={"host": "0.0.0.0"})

while True:
    time.sleep(1)
```

**Translator.py**
```
from starlette.requests import Request
import ray
from ray import serve

app = FastAPI()

@serve.deployment(name="main", route_prefix="/")
@serve.ingress(app)
class Translator:
    def __init__(self):
        pass
      
    async def __call__(self, http_request: Request) -> str:
        return "hello"

```

default_worker.py crashes with output:
```
2023-11-30 00:01:26,715 DEBUG worker.py:646 -- Retaining 16310a0f0a45af5cb10accb0881646893aef9f9f0100000001000000
(raylet) bash: --startup-token=16: command not found
(raylet) usage: default_worker.py [-h] --cluster-id CLUSTER_ID --node-ip-address
(raylet)                          NODE_IP_ADDRESS --node-manager-port NODE_MANAGER_PORT
(raylet)                          [--raylet-ip-address RAYLET_IP_ADDRESS]
(raylet)                          --redis-address REDIS_ADDRESS --gcs-address
(raylet)                          GCS_ADDRESS [--redis-password REDIS_PASSWORD]
(raylet)                          --object-store-name OBJECT_STORE_NAME
(raylet)                          [--raylet-name RAYLET_NAME]
(raylet)                          [--logging-level {debug,info,warning,error,critical}]
(raylet)                          [--logging-format LOGGING_FORMAT]
(raylet)                          [--temp-dir TEMP_DIR] [--storage STORAGE]
(raylet)                          [--load-code-from-local] [--worker-type WORKER_TYPE]
(raylet)                          --metrics-agent-port METRICS_AGENT_PORT
(raylet)                          --runtime-env-agent-port RUNTIME_ENV_AGENT_PORT
(raylet)                          [--object-spilling-config OBJECT_SPILLING_CONFIG]
(raylet)                          [--logging-rotate-bytes LOGGING_ROTATE_BYTES]
(raylet)                          [--logging-rotate-backup-count LOGGING_ROTATE_BACKUP_COUNT]
(raylet)                          [--runtime-env-hash RUNTIME_ENV_HASH] --startup-token
(raylet)                          STARTUP_TOKEN [--ray-debugger-external]
(raylet)                          [--session-name SESSION_NAME] [--webui WEBUI]
(raylet)                          --worker-launch-time-ms WORKER_LAUNCH_TIME_MS
(raylet)                          [--worker-preload-modules WORKER_PRELOAD_MODULES]
(raylet) default_worker.py: error: the following arguments are required: --startup-token, --worker-launch-time-ms
```

I investigated using strace and got arguments which are passed to default_worker.py:
```
[pid 17910] +++ exited with 0 +++
[pid 17815] execve("/usr/bin/bash", ["bash", "-c", "exec /home/ray/anaconda3/bin/python /home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/workers/default_worker.py --node-ip-address=192.168.0.2 --node-manager-port=42547 --object-store-name=/tmp/ray/session_2023-12-01_13-42-35_650605_16416/sockets/plasma_store --raylet-name=/tmp/ray/session_2023-12-01_13-42-35_650605_16416/sockets/raylet --redis-address=None --temp-dir=/tmp/ray --metrics-agent-port=59796 --runtime-env-agent-port=62111 --logging-rotate-bytes=536870912 --logging-rotate-backup-count=5 --runtime-env-agent-port=62111 --gcs-address=192.168.0.2:6379 --session-name=session_2023-12-01_13-42-35_650605_16416 --temp-dir=/tmp/ray --webui=192.168.0.2:8265 --cluster-id=e0269812994822291b61b2be7bc4a045f747c265f98712c687ce2dc3 --storage=s3://explore/ray?&endpoint_override=http://192.168.0.4:9000 --startup-token=16 --worker-launch-time-ms=1701466993176 --runtime-env-hash=1559810005"], 0x101b950 /* 71 vars */) = 0
strace: Process 17922 attached
[pid 17922] execve("/home/ray/anaconda3/bin/python", ["/home/ray/anaconda3/bin/python", "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/workers/default_worker.py", "--node-ip-address=192.168.0.2", "--node-manager-port=42547", "--object-store-name=/tmp/ray/session_2023-12-01_13-42-35_650605_16416/sockets/plasma_store", "--raylet-name=/tmp/ray/session_2023-12-01_13-42-35_650605_16416/sockets/raylet", "--redis-address=None", "--temp-dir=/tmp/ray", "--metrics-agent-port=59796", "--runtime-env-agent-port=62111", "--logging-rotate-bytes=536870912", "--logging-rotate-backup-count=5", "--runtime-env-agent-port=62111", "--gcs-address=192.168.0.2:6379", "--session-name=session_2023-12-01_13-42-35_650605_16416", "--temp-dir=/tmp/ray", "--webui=192.168.0.2:8265", "--cluster-id=e0269812994822291b61b2be7bc4a045f747c265f98712c687ce2dc3", "--storage=s3://explore/ray?"], 0x561ebe4434f0 /* 71 vars */strace: Process 17923 attached
) = 0
[pid 17923] +++ exited with 127 +++

```

It seems that equal sign in  --storage="s3://explore/ray?&endpoint_override**=**http://192.168.0.4:9000"` parameter is not escaped before passing to default_worker.py and it causes  that

### Versions / Dependencies

Tested ray version 2.7.1  and 2.8.1

### Reproduction script

I provided all scripts above

### Issue Severity

High: It blocks me from completing my task.

**Comment by tvildo** [comment]
Bug seems to be in worker_pool.cc here 

https://github.com/ray-project/ray/blob/f07e319ac7ea3ec3764599a519f59db3f55c8e8c/src/ray/raylet/worker_pool.cc#L299-L319

**Comment by jjyao** [comment]
@tvildo do you want to create a PR for it? We are happy to review it.

**Comment by jjyao** [comment]
Fixed the issue for macos and linux with #42332.

Keep this issue open to track the windows fix. @mattip could you help with the Windows fix. I tried but don't know how to correctly escape the command. After the windows fix, we should be able to unskip the test for windows.

**Comment by anyscalesam** [comment]
@jjyao can you please review and merge @mattip PR; you are listed as reviewer.
Matti as discused please cut a separate ticket with labels core, windows, p1, enhancement for solving for the holsitic case of all special characters in http urls as discussed

here is allof them https://chat.openai.com/share/4fd49922-f27a-4169-af91-2a7345971609
